go/worker/compute/executor/committee/node.go

```diff
@@ -188,22 +214,65 @@ func (n *Node) getMetricLabels() prometheus.Labels {
 
 // HandlePeerMessage implements NodeHooks.
 func (n *Node) HandlePeerMessage(ctx context.Context, message *p2p.Message) (bool, error) {
-	if message.TxnSchedulerBatch != nil {
+	n.logger.Debug("received peer message", "message", message)
+
+	switch {
+	case message.Tx != nil:
+		call := message.Tx.Data
+
+		// Check if we are a leader. Note that we may be in the middle of a
+		// transition, but this shouldn't matter as the client will retry.
+		n.commonNode.CrossNode.Lock()
+		height := n.commonNode.CurrentHeight
+		n.commonNode.CrossNode.Unlock()
+
+		// TODO: can accept it even if we're not.
+		if !n.commonNode.Group.GetEpochSnapshot().IsTransactionSchedulerLeader(height) {
+			n.logger.Error("unable to handle transaction message, not transaction scheduling leader",
+				"err", errNotTxnSchedLeader,
+				"current_epoch", n.commonNode.Group.GetEpochSnapshot().GetEpochNumber(),
+				"current_height", height,
+			)
+			return true, nil
+		}
+
+		if n.scheduleCheckTxEnabled {
+			// Check transaction before queuing it.
+			if err := n.checkTx(ctx, call); err != nil {
+				return true, p2pError.Permanent(err)
+			}
+			n.logger.Debug("worker CheckTx successful, queuing transaction")
+		}
+
+		err := n.QueueTx(call)
+		if err != nil {
+			n.logger.Error("unable to Queue transaction",
+				"err", err,
+			)
+			return true, p2pError.Permanent(err)
+		}
+		return true, nil
+
+	case message.TxnSchedulerBatch != nil:
 		crash.Here(crashPointBatchReceiveAfter)
 
 		sbd := message.TxnSchedulerBatch
 
+		epoch := n.commonNode.Group.GetEpochSnapshot()
+
 		// Before opening the signed dispatch message, verify that it was
 		// actually signed by the current transaction scheduler.
-		epoch := n.commonNode.Group.GetEpochSnapshot()
-		txsc := epoch.GetTransactionSchedulerCommittee()
-		if !txsc.PublicKeys[sbd.Signature.PublicKey] {
+		n.commonNode.CrossNode.Lock()
+		height := n.commonNode.CurrentHeight
+		n.commonNode.CrossNode.Unlock()
+		txsc := epoch.GetTransactionSchedulerLeader(height)
+		if !txsc.PublicKey.Equal(sbd.Signature.PublicKey) {
```

Comment:
This should probably use `epoch.VerifyTxnSchedulerSignature`.

Reply:
Also for the height this needs to take into account the current "round height" (if any), similar to what the pool does. The easiest way to do that is to first update `ExecutorCommittedEvent` to include a `CurrentRoundHeight` and then make sure that the executor node handles the `ExecutorCommittedEvent` event in `HandleNewEventLocked` and updates its view of the round height. If there is no current round height, it should use the current height.